File: backend/app/schemas/finalization.py
"""Schemas for Step 5 — final protocol summary, confirmation & closing."""
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.enums import (
    AppointmentType,
    PatientSex,
    RedFlagSeverity,
    SessionStatus,
    TreatmentItemKind,
)
from app.schemas.protocol import ProtocolOut
from app.schemas.session import SessionOut


class DoctorInfo(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    full_name: str


class SelectedDiagnosisInfo(BaseModel):
    title: str
    icd10_code: str | None = None


class RedFlagSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    label: str
    description: str | None
    severity: RedFlagSeverity | None
    acknowledged_at: datetime | None
    doctor_note: str | None


class TreatmentItemSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    kind: TreatmentItemKind
    title: str
    details: str | None
    dosage: str | None
    duration: str | None
    order_index: int
    is_confirmed: bool


class SessionSummaryBlock(BaseModel):
    """A subset of ``SessionOut`` used inside the summary card."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    doctor_id: uuid.UUID
    patient_full_name: str
    patient_age: int
    patient_sex: PatientSex
    appointment_type: AppointmentType
    status: SessionStatus
    consent_given_at: datetime | None
    created_at: datetime
    updated_at: datetime


class SessionSummaryOut(BaseModel):
    session: SessionSummaryBlock
    doctor: DoctorInfo
    protocol: ProtocolOut | None
    selected_diagnosis: SelectedDiagnosisInfo | None
    red_flags: list[RedFlagSummary]
    treatment_plan_items: list[TreatmentItemSummary]


class ConfirmSessionOut(BaseModel):
    session: SessionOut
    protocol: ProtocolOut


class CloseSessionOut(BaseModel):
    session: SessionOut

File: frontend/src/api/finalization.ts
import { apiClient } from './client'
import type {
  Protocol,
  RedFlagSeverity,
  TreatmentKind,
} from './analysis'
import type { AppointmentSession } from './sessions'

export interface DoctorInfo {
  id: string
  full_name: string
}

export interface SelectedDiagnosisInfo {
  title: string
  icd10_code: string | null
}

export interface RedFlagSummary {
  id: string
  label: string
  description: string | null
  severity: RedFlagSeverity | null
  acknowledged_at: string | null
  doctor_note: string | null
}

export interface TreatmentItemSummary {
  id: string
  kind: TreatmentKind
  title: string
  details: string | null
  dosage: string | null
  duration: string | null
  order_index: number
  is_confirmed: boolean
}

export interface SessionSummary {
  session: AppointmentSession
  doctor: DoctorInfo
  protocol: Protocol | null
  selected_diagnosis: SelectedDiagnosisInfo | null
  red_flags: RedFlagSummary[]
  treatment_plan_items: TreatmentItemSummary[]
}

export async function getSessionSummary(
  id: string,
): Promise<SessionSummary> {
  const { data } = await apiClient.get<SessionSummary>(
    `/sessions/${id}/summary`,
  )
  return data
}

export interface ConfirmSessionResponse {
  session: AppointmentSession
  protocol: Protocol
}

export async function confirmSession(
  id: string,
): Promise<ConfirmSessionResponse> {
  const { data } = await apiClient.post<ConfirmSessionResponse>(
    `/sessions/${id}/confirm`,
  )
  return data
}

export interface CloseSessionResponse {
  session: AppointmentSession
}

export async function closeSession(
  id: string,
): Promise<CloseSessionResponse> {
  const { data } = await apiClient.post<CloseSessionResponse>(
    `/sessions/${id}/close`,
  )
  return data
}

/** Downloads the protocol PDF as a binary blob and triggers a browser save. */
export async function downloadSessionPdf(
  id: string,
  filename?: string,
): Promise<void> {
  const resp = await apiClient.get(`/sessions/${id}/pdf`, {
    responseType: 'blob',
  })
  const blob = new Blob([resp.data], { type: 'application/pdf' })
  const url = URL.createObjectURL(blob)
  const link = document.createElement('a')
  link.href = url
  link.download = filename ?? `protocol_${id}.pdf`
  document.body.appendChild(link)
  link.click()
  document.body.removeChild(link)
  // Defer revoke so the browser finishes triggering the download.
  setTimeout(() => URL.revokeObjectURL(url), 1000)
}
